pkg/reconciler/ingress/status.go

```diff
@@ -0,0 +1,268 @@
+// Also, it means that IsReady is not called concurrently.
+func (m *StatusManager) IsReady(vs *v1alpha3.VirtualService) (bool, error) {
+	key := fmt.Sprintf("%s/%s", vs.Namespace, vs.Name)
+
+	// Find the probe host
+	var probeHost string
+	for _, host := range vs.Spec.Hosts {
+		if strings.HasSuffix(host, resources.ProbeHostSuffix) {
+			if probeHost != "" {
+				return false, fmt.Errorf("only one probe host can be defined in a VirtualService, found at least 2: %s, %s", probeHost, host)
+			}
+			probeHost = host
+		}
+	}
+	if probeHost == "" {
+		return false, fmt.Errorf("only a VirtualService with a probe can be probed. Hosts: %v", vs.Spec.Hosts)
+	}
+
+	m.mu.Lock()
+	if state, ok := m.probingStates[key]; ok {
+		if state.probeHost == probeHost {
+			m.mu.Unlock()
+			state.lastAccessed = time.Now()
+			return atomic.LoadInt32(&state.pendingCount) == 0, nil
+		}
+
+		// Cancel the polling for the outdated version
+		state.cancel()
+		delete(m.probingStates, key)
+	}
+	m.mu.Unlock()
+
+	podIPs, err := m.listVirtualServicePodIPs(vs)
+	if err != nil {
+		return false, fmt.Errorf("failed to list the IP addresses of the Pods impacted by the VirtualService: %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	state := &probingState{
+		virtualService: vs,
+		probeHost:      probeHost,
+		pendingCount:   int32(len(podIPs)),
+		lastAccessed:   time.Now(),
+		context:        ctx,
+		cancel:         cancel,
+	}
+
+	func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		m.probingStates[key] = state
+	}()
+	for _, podIP := range podIPs {
+		workItem := &workItem{
+			probingState: state,
+			podIP:        podIP,
+		}
+		m.workQueue.AddRateLimited(workItem)
+	}
+
+	return len(podIPs) == 0, nil
+}
+
+// Start starts the StatusManager background operations
+func (m *StatusManager) Start(done <-chan struct{}) {
+	// Start the worker goroutines
+	for i := 0; i < probeConcurrency; i++ {
+		go func() {
+			for m.processWorkItem() {
+			}
+		}()
+	}
+
+	// Cleanup the states periodically
+	go func() {
+		wait.Until(m.expireOldStates, time.Minute, done)
+	}()
+
+	// Stop processing the queue when cancelled
+	go func() {
+		<-done
+		m.workQueue.ShutDown()
+	}()
+}
+
+// expireOldStates removes the states that haven't been accessed in a while.
+func (m *StatusManager) expireOldStates() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for key, state := range m.probingStates {
+		if time.Since(state.lastAccessed) > stateExpiration {
+			state.cancel()
+			delete(m.probingStates, key)
+		}
+	}
+}
+
+// processWorkItem processes a single work item from workQueue.
+// It returns false when there is no more items to process, true otherwise.
+func (m *StatusManager) processWorkItem() bool {
+	obj, shutdown := m.workQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	defer m.workQueue.Done(obj)
+
+	// Discard the item if it is not of the expected type
+	item, ok := obj.(*workItem)
+	if !ok {
+		m.workQueue.Forget(obj)
```

Comment:
log? this should not really happen

Reply:
Shouldn't this be a panic? As you said, it should really not happen.
On the other hand, it can be safely ignored.

Reply:
log.Fatal it is.